frontend/catalyst/utils/toml.py

```diff
@@ -115,57 +164,178 @@ def get_gates(config: TOMLDocument, path: List[str], shots_present: bool) -> Dic
+    config: TOMLDocument, program_features: ProgramFeatures
+) -> Dict[str, dict]:
     """Get gates that will be decomposed according to PL's decomposition rules.
 
     Args:
         config (TOMLDocument): Configuration dictionary
     """
     schema = int(config["schema"])
     if schema == 1:
-        return get_gates(config, ["operators", "gates", 0, "decomp"], shots_present)
+        return get_gates(config, ["operators", "gates", 0, "decomp"], program_features)
     elif schema == 2:
-        return get_gates(config, ["operators", "gates", "decomp"], shots_present)
+        return get_gates(config, ["operators", "gates", "decomp"], program_features)
 
     raise CompileError(f"Unsupported config schema {schema}")
 
 
-def get_matrix_decomposable_gates(config: TOMLDocument, shots_present: bool) -> Dict[str, dict]:
+def get_matrix_decomposable_gates(
+    config: TOMLDocument, program_features: ProgramFeatures
+) -> Dict[str, dict]:
     """Get gates that will be decomposed to QubitUnitary.
 
     Args:
         config (TOMLDocument): Configuration dictionary
     """
     schema = int(config["schema"])
     if schema == 1:
-        return get_gates(config, ["operators", "gates", 0, "matrix"], shots_present)
+        return get_gates(config, ["operators", "gates", 0, "matrix"], program_features)
     elif schema == 2:
-        return get_gates(config, ["operators", "gates", "matrix"], shots_present)
+        return get_gates(config, ["operators", "gates", "matrix"], program_features)
 
     raise CompileError(f"Unsupported config schema {schema}")
+
+
+def get_operation_properties(config_props: dict) -> OperationProperties:
+    """Load operation properties from config"""
+    properties = config_props.get("properties", {})
+    return OperationProperties(
+        invertible="invertible" in properties,
+        controllable="controllable" in properties,
+        differentiable="differentiable" in properties,
+    )
+
+
+def get_device_config(
+    config: TOMLDocument, program_features: ProgramFeatures, device_name: str
+) -> DeviceConfig:
+    """Load TOML document into the DeviceConfig structure"""
+    # pylint: disable=too-many-branches
+
+    schema = int(config["schema"])
+
+    native_gate_props = {}
+    for g, props in get_native_gates(config, program_features).items():
+        native_gate_props[g] = get_operation_properties(props)
+
+    matrix_decomp_props = {}
+    for g, props in get_matrix_decomposable_gates(config, program_features).items():
+        matrix_decomp_props[g] = get_operation_properties(props)
+
+    decomp_props = {}
+    for g, props in get_decomposable_gates(config, program_features).items():
+        decomp_props[g] = get_operation_properties(props)
+
+    observable_props = {}
+    for g, props in get_observables(config, program_features).items():
+        observable_props[g] = get_operation_properties(props)
+
+    if schema == 1:
+        # TODO: remove after PR #642 is merged in lightning
```

Comment:
Can you split the schema 1 as a separate function?

Reply:
Done.